streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala

```diff
@@ -82,13 +113,23 @@ class ReceiverTracker(ssc: StreamingContext, skipReceiverLaunch: Boolean = false
         "ReceiverTracker", new ReceiverTrackerEndpoint(ssc.env.rpcEnv))
       if (!skipReceiverLaunch) receiverExecutor.start()
       logInfo("ReceiverTracker started")
+      trackerState = Started
     }
   }
 
   /** Stop the receiver execution thread. */
   def stop(graceful: Boolean): Unit = synchronized {
-    if (!receiverInputStreams.isEmpty && endpoint != null) {
+    if (isTrackerStarted) {
       // First, stop the receivers
+      // acquire "stoppingTrackerLock" so that setting trackerState to "Stopping" and registering
+      // receivers won't happen at the same time
+      stoppingTrackerLock.synchronized {
+        trackerState = Stopping
+        if (!skipReceiverLaunch) {
+          // Send the stop signal to all the receivers
+          receiverExecutor.stopReceivers()
```

Comment:
I find this weird, to call stop `receiverExecutor.stopReceivers()` and then again `receiverExecutor.stop()`
Its breaks the simple abstraction that all the `receiverTracker` has to know about `receiverExecutor`'s internal `stopReceiver` as well as `stop`.